gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java

```diff
@@ -160,26 +161,69 @@ public HiveRegistrationPolicyBase(State props) throws IOException {
+   *  2. Table names specified by <code>additional.hive.table.names</code>
+   *
+   * In table names above, the {@value PRIMARY_TABLE_TOKEN} if present is also replaced by the
+   * table name obtained via {@link #getTableName(Path)}.
+   *
+   * @param path Path for the table on filesystem.
+   * @return Table names to register.
    */
   protected Iterable<String> getTableNames(Path path) {
+    List<String> tableNames = getTableNames(Optional.<String>absent(), path);
+
+    Preconditions.checkState(!tableNames.isEmpty(), "Hive table name not specified");
+    return tableNames;
+  }
+
+  /***
+   * Obtain Hive table names filtered by <code>dbPropertyFilter</code> (if present).
+   *
+   * The returned {@link List} contains:
+   *  A. If <code>dbPropertyFilter</code> is absent:
+   *    1. Table name returned by {@link #getTableName(Path)}
+   *    2. Table names specified by <code>additional.hive.table.names</code>
+   *  B. If dbPropertyFilter is present:
+   *    1. Table names specified by <code>dbPropertyFilter.additional.hive.table.names</code>
+   *
+   * In table names above, the {@value PRIMARY_TABLE_TOKEN} if present is also replaced by the
+   * table name obtained via {@link #getTableName(Path)}.
+   *
+   * @param dbPropertyFilter Prefix to the property <code>additional.table.names</code>, to obtain table names only
+   *                         for the specified db. Eg. If <code>dbPropertyFilter</code> is db, then
+   *                         <code>db.additional.table.names</code> is the resolved property name.
+   * @param path Path for the table on filesystem.
+   * @return Table names to register.
+   */
+  protected List<String> getTableNames(Optional<String> dbPropertyFilter, Path path) {
     List<String> tableNames = Lists.newArrayList();
 
-    Optional<String> tableName;
-    if ((tableName = getTableName(path)).isPresent()) {
-      tableNames.add(tableName.get());
+    Optional<String> primaryTableName;
+    if ((primaryTableName = getTableName(path)).isPresent() && !dbPropertyFilter.isPresent()) {
+      tableNames.add(primaryTableName.get());
+    }
+
+    String additionalNamesProp;
+    if (dbPropertyFilter.isPresent()) {
+      additionalNamesProp = String.format("%s.%s", dbPropertyFilter.get(), ADDITIONAL_HIVE_TABLE_NAMES);
+    } else {
+      additionalNamesProp = ADDITIONAL_HIVE_TABLE_NAMES;
     }
 
-    if (!Strings.isNullOrEmpty(this.props.getProp(ADDITIONAL_HIVE_TABLE_NAMES))) {
-      for (String additionalTableName : this.props.getPropAsList(ADDITIONAL_HIVE_TABLE_NAMES)) {
-        tableNames.add(this.tableNamePrefix + additionalTableName + this.tableNameSuffix);
+    if (!Strings.isNullOrEmpty(this.props.getProp(additionalNamesProp))) {
+      for (String additionalTableName : this.props.getPropAsList(additionalNamesProp)) {
+        String resolvedTableName = primaryTableName.isPresent() ?
+            StringUtils.replace(additionalTableName, PRIMARY_TABLE_TOKEN, primaryTableName.get()) :
```

Comment:
Why is the property `$PRIMARY_TABLE` token required? Can we not use `${hive.table.name}` ?